src/input.cpp

```diff
@@ -1189,6 +1189,25 @@ const std::string &input_context::handle_input( const int timeout )
     next_action.type = input_event_t::error;
     const std::string *result = &CATA_ERROR;
     while( true ) {
+
+        //Allows "toggle_language_to_en" to also work on contexts other than "DEFAULTMODE"
+        const std::string pressed_key = next_action.text;
+        if( !pressed_key.empty() ) {
+            action_attributes attributes = inp_mngr.get_action_attributes( "toggle_language_to_en",
+                                           "DEFAULTMODE" );
+            if( !attributes.input_events.empty() ) {
+                input_event input = attributes.input_events.front();
+                if( !input.sequence.empty() ) {
+                    char keybind = input.sequence.front();
+                    if( *pressed_key.c_str() == keybind ) {
+                        g->toggle_language_to_en();
+                        g->invalidate_main_ui_adaptor();
+                        ui_manager::redraw_invalidated();
+                    }
+                }
+            }
+        } //
```

Comment:
This feels too intrusive to me. It uses a keybind that is not registered to the specific context so there might be conflicts and people cannot reassign it for the specific context, unlike all other actions.

It might be better to make it work similar to "HELP_KEYBINDINGS", and you can probably register it by default if you don't feel like registering it manually everywhere `input_context` is used.

Also, comparing `text` with the hotkey is not guaranteed to work because some keys do not have a corresponding character, and in key code mode there will not be any text, but only keys. You should compare the action string instead.

Furthermore, currently only the main UI is invalidated, which I think happens to invalidate all UIs when a save is loaded because the main UI uses the whole screen area, but it might not work when a save is not loaded (for example, in the main menu) if I'm not mistaken.

Also, I think some UIs are storing the translated text so toggling the language won't work in these UIs, but that can be addressed in separate PRs.

Reply:
The fact the key is not registered is incorrect, the keybind is indeed registered to a specific context, which is DEFAULTMODE - what the code above is doing is extending it's functionality across context, because if you have a function that quickly swap languages to either help you translate the game or learn another language, the last thing you want is to having to exit the item description menu, swap language, try to find the item and access again it's description and doing that back and forth - it wouldn't be user friendly at all, so the fact this key works globally and is not meant to be re-assigned is actually a feature of this functionality. Also, the key starts unbound.

I had attempted messing with HELP_KEYBINDINGS and all attempted have failed, so unless I'm shown exactly how, I can't really achieve an implementation better than this.
Attempting registering the same key to every single context I don't think it would work either because then you would have several "Toggle language to english" in the game's keybind option menu, and there is meant to be only one entry of that that controls it across the game.

Changing it to compare the action I can do! :)

Regarding some ui not updating certain elements (like in the inventory ui) I'm currently working on that, I was pointed towards "inventory_entry::make_entry_cell_cache()" and "Prepare_layout()" so I'll be checking into it today

Reply:
I think you can remove the `"DEFAULTMODE"` context so it becomes a global keybinding. Then you can register it in the constructors of `input_context` so every input context has it by default. Then you can handle it similar to how the code handles `"HELP_KEYBINDINGS"` below. The player can then register a global hotkey for the action using `=`, so it won't make them have to register it again unless they want to solve a conflict, and it will only be shown once in the keybindings menu, but in every input context so it can be rebound whenever you want.

Reply:
> I think you can remove the "DEFAULTMODE" context so it becomes a global keybinding. 
If I remove the "ctxt.register_action( "toggle_language_to_en" );" under the DEFAULTMODE" context, then there is not an entry under keybind for this functionality. 

>What do you think about the UI invalidation problem I mentioned? Does it currently work in main menu, world creation UI, etc?
The main ui when you start the game has mixed results - The main menu itself (as soon as you launch the game) is not updating, while in the world generations menu and similar ui on top of the starting menu, it works but there you need also to move the mouse after you press the "toggle" key to make it refresh/redraw, it seems.
If you guys can tell me how to refresh that manually would be great, but honestly, even if this ended up not being solved I don't think it would be a reason to lose sleep on it, since the 99% of the reading is performed inside inventory and item description, that is where I'm focusing my attention first.  Currently, they key responds inside the inventory but not from inside an "item info" panel so that needs figuring out

Reply:
I mean removing the `"category": "DEFAULTMODE"` line from `keybindings.json`. An action without a category takes effect in all input contexts.

And if you want to invalidate all UIs, you can create a new function in `ui_manager.cpp` that sets the `invalidated` flag of all `ui_adaptor`s on `ui_stack` and call `invalidation_consistency_and_optimization` (i.e. similar to `ui_adaptor::invalidate` but invalidates all UIs). Then you can call that function when switching language so all UIs will be refreshed on the next redraw.

Reply:
I've attempted removing the "category" entry from the json, it makes it disappear from the keybinds menu :\

Reply:
Maybe is because I've also removed "ctxt.register_action( "toggle_language_to_en" );" using the "DEFAULTMODE" context.  
You said "Then you can register it in the constructors of input_context so every input context has it by default.", maybe that's the crucial step I'm missing, but I don't know where exactly I should do that because the input_context constructor is very strange, and what context I should use to register this

Reply:
ok, I've attempted using the function register_action("toggle_language_to_en"); in what it seemed to me the point you've suggested, and it worked - although I'm starting a new branch because otherwise I'm bound to mix things up with the way it currently works, and if it fails I don't want to lose current progress x_x
Hope is ok if I'll be asking further questions

Reply:
Can you go a bit into detail about what do you mean with "**Then you can handle it similar to how the code handles "HELP_KEYBINDINGS" below.** "

Reply:
I will try :D

Reply:
I think I've followed your direction correctly, although unfortunately the starting menu is still not getting refreshed :\

Reply:
It is all strings not changing or some strings not changing? The main menu stores some of its translated strings so these are expected to not change.

Reply:
The only strings changing are the one showing the Version of the game, the one that says "Bugs?  Suggestions?  Use links in MOTD to report them." and the string saying "Tip of the day:" but not the tip itself.   Everything else stays unchanged.
But honestly, that's not a big, since this feature is mostly used/useful in inventory and item descriptions.  Currently with some help I've managed to make all the item strings in the inventory re-cache (is that the term?) so inventory is getting fully updated when swapping language, but I haven't figured out yet a good way to do this for the menu with the item description itself ( should be the one associated to the draw_item_info() function )  -  if you have any suggestion for that it would be much appreciated and likely would make this PR ready for merge ^_^

Reply:
Yeah, that's what I expected for the main menu because the other strings are stored as translated strings when entering the menu.

The item info also stores the translated strings so you probably need to regenerate the item info as a whole. I'd recommend adding a new overload of `draw_item_info` that takes a `std::function<item_info_data()>` instead of a `item_info_data`, and call that function at the start and when the language changes to get the updated item info data.

Reply:
thank you for the suggestion, although I'm seeying it only now and in the meantime I was solving this differently, I've simply moved the code that draws the inventory_item_info panel into the loop below that code so that it can be executed again if the language changes

Reply:
I was thinking about the item info panel opened by pressing `e` inside the invetory UI, which I think has its own input loop. Does that also work?